src/main/ruby/core/rbconfig.rb

```diff
@@ -166,7 +166,7 @@ module RbConfig
   cc             = CLANG
   cpp            = cc
   linkflags      = ['-g',                                     # Show debug information such as line numbers in backtrace
-                    '-Werror=implicit-function-declaration',  # To make missing C ext functions very clear
+                    '-Wimplicit-function-declaration',        # To make missing C ext functions clear
```

Comment:
I'm not convinced this is better except for of course for compatibility. It can create hard to debug situations where the only tool is `strace` + knowing the syscall and what kind of arguments it should receive.

Maybe we should have different sets of flags for internal cexts and gem cexts.
We likely want to be strict with internal cexts (most importantly what goes in ruby.su), but also permissive for gem cexts.
I don't see an easy way to do that though.

Does the warning appear when compiling our own cexts? What about when compiling a gem? It goes to the mkmf.log?

Reply:
> It can create hard to debug situations where the only tool is strace + knowing the syscall and what kind of arguments it should receive

Yes I remember now - the problem wasn't just missing declarations - it was declarations with the wrong type, and then things only going wrong very late on when it was hard to debug.

What about a debug flag for development?

Reply:
The warning does appear in mkmf.log. I haven't audited all our C extensions. CI passed and this fixed the immediate problem I had with the `bson` gem.

I'm okay having a different set of flags for development, but I think our default needs to be permissive to match MRI.

Reply:
For our own cexts it shows up as:
```
ruby.c:39:9: warning: implicit declaration of function 'gethostname2' is invalid in C99
      [-Wimplicit-function-declaration]
        gethostname2(foo, 24);
        ^
1 warning generated.
linking shared-object ruby.su
```
on `jt build`.
That's good enough for me.